llvm/lib/IR/LLVMContext.cpp

```diff
@@ -256,10 +256,13 @@ void LLVMContext::diagnose(const DiagnosticInfo &DI) {
       RS->emit(*OptDiagBase);
 
   // If there is a report handler, use it.
-  if (pImpl->DiagHandler &&
-      (!pImpl->RespectDiagnosticFilters || isDiagnosticEnabled(DI)) &&
-      pImpl->DiagHandler->handleDiagnostics(DI))
-    return;
+  if (pImpl->DiagHandler) {
+    if (DI.getSeverity() == DS_Error)
+      pImpl->DiagHandler->HasErrors = true;
```

Comment:
This unfortunately exposes the detail of DiagnosticHandler, but the alternative (set HasErrors in `DiagnosticHandler::handleDiagnostics`) isn't better and a new client may forget to call the base.

Reply:
Perhaps move the set of HasErrors into a new non-virtual method that we call below just before calling handleDiagnostics.

Also, should this be set when handleDiagnostics is not called?

Reply:
I've considered this choice. Currently,

* LLVMContext::diagnose and handleDiagnostics need to cooperate and handle part of the process.
* handleDiagnostics is only called by LLVMContext::diagnose.

If we call `handleDiagnostics` in more places, it's clear that we need to duplicate some code in `diagnose` to the new call site, at least the `LLVMRemarkStreamer` stuff. Therefore, adding a new method to DiagnosticHandler does not abstract away things, and do not make things less error-prone.

Reply:
> Also, should this be set when handleDiagnostics is not called?

The `RespectDiagnosticFilters` condition is here to suppress remarks for legacy LTOCodeGenerator (test: `llvm/test/LTO/X86/diagnostic-handler-remarks.ll`). It's not used for errors. I think conveying the `HasErrors` bit seems more useful, although the client does not use this information.

Reply:
I didn't mean calling handle Diagnostics in more places. I just meant rather than directly setting the HasErrors flags here, do that in a new non-virtual method (e.g. prepareToHandleDiagnostics() or whatever), and call it here just before calling handle Diagnostics. To abstract away what it is actually doing and leave the setting of the flag to the DiagnosticHandler class.

Reply:
Since `DiagnosticHandler` members are all public (and I do not want to change this), it seems that we could just avoid defining a new function, which is not supposed to be called elsewhere anyway...